vowpalwabbit/v_array.h

```diff
@@ -39,24 +54,71 @@ struct v_array
   T* end_array;
   size_t erase_count;
 
-  using iterator = T*;
+  using value_type = T;
+  using reference = value_type&;
+  using const_reference = const value_type&;
+  using pointer = value_type*;
+  using const_pointer = const value_type*;
+  using iterator = value_type*;
+  using const_iterator = const value_type*;
 
   // enable C++ 11 for loops
-  inline T*& begin() { return _begin; }
-  inline T*& end() { return _end; }
+  inline iterator& begin() { return _begin; }
+  inline iterator& end() { return _end; }
+
+  inline const_iterator begin() const { return _begin; }
+  inline const_iterator end() const { return _end; }
 
-  inline const T* begin() const { return _begin; }
-  inline const T* end() const { return _end; }
+  inline const_iterator cbegin() const { return _begin; }
+  inline const_iterator cend() const { return _end; }
 
-  inline T* cbegin() const { return _begin; }
-  inline T* cend() const { return _end; }
+  v_array() noexcept : _begin(nullptr), _end(nullptr), end_array(nullptr), erase_count(0) {}
+  ~v_array() { delete_v_array(); }
+
+  v_array(v_array<T>&& other) noexcept
+  {
+    erase_count = 0;
+    _begin = nullptr;
+    _end = nullptr;
+    end_array = nullptr;
+
+    std::swap(_begin, other._begin);
+    std::swap(_end, other._end);
+    std::swap(end_array, other.end_array);
+    std::swap(erase_count, other.erase_count);
+  }
+
+  v_array<T>& operator=(v_array<T>&& other) noexcept
+  {
+    delete_v_array();
```

Comment:
Is the deletion necessary here? The source object would theoretically clean everything up once it goes out of scope

Reply:
The moved out of object doesn't necessarily go out of scope though. Someone could use it. It seems better to leave it in this safe and well defined state.

Reply:
Will change to just a swap